File: C/1.工厂模式/src/ShapeFactory.c
/**
 * @file ShapeFactory.c
 * @author {Layne} ({shu_huanglei@163.com})
 * @brief
 * @version 0.1
 * @date 2020-07-16
 *
 * @copyright Copyright (c) 2020
 *
 */
#include "ShapeFactory.h"

#include <stdlib.h>
#include <string.h>

#include "ShapeStruct.h"

extern struct Shape* circle_create(void);
extern struct Shape* rectangle_create(void);
extern struct Shape* square_create(void);

Shape* shape_factory_get_shape(const char* shape_type) {
    if (NULL == shape_type) return NULL;

    if (0 == strcasecmp("CIRCLE", shape_type)) {
        return circle_create();
    } else if (0 == strcasecmp("RECTANGLE", shape_type)) {
        return rectangle_create();
    } else if (0 == strcasecmp("SQUARE", shape_type)) {
        return square_create();
    }

    return NULL;
}

File: Objective-C/1.FactoryPattern/ShapeFactory.m
//
//  ShapeFactory.m
//  1.FactoryPattern
//
//  Created by Layne on 2020/10/17.
//

#import "ShapeFactory.h"
#import "Circle.h"
#import "Rectangle.h"
#import "Square.h"

@implementation ShapeFactory

+ (id<ShapeProtocol>)getShape:(NSString*)shapeType {
    if (shapeType == nil) {
        return nil;
    }
    if ([shapeType isEqualToString:@"CIRCLE"]) {
        return [[Circle alloc] init];
    } else if ([shapeType isEqualToString:@"RECTANGLE"]) {
        return [[Rectangle alloc] init];
    } else if ([shapeType isEqualToString:@"SQUARE"]) {
        return [[Square alloc] init];
    }
    return nil;
}

@end
